### PR TITLE
docs: Formatters page updates

### LINKED ISSUE #16476: Formatters Page Clean Up & Expansion
### ESLint version

v8.x

### What problem do you want to solve?

clean up copy of the Formatters page, and expand key aspects to create a better reader experience.

### What do you think is the correct solution?

Miscellaneous changes on [Formatters](https://eslint.org/docs/latest/user-guide/formatters/) page: 
- In the example source section, make fullOfProblems.js and .eslintrc: not headers. 
- For the Example Outputs section, provide an example of the CLI input, and also give a few sentence description of what each of the output is, and why you might want to use it.
- Other small cosmetic changes and refactors

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by bpmutter**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Miscellaneous changes on [Formatters](https://eslint.org/docs/latest/user-guide/formatters/) page:

- In the example source section, make fullOfProblems.js and .eslintrc: not headers.
- For the Example Outputs section, provide an example of the CLI input, and also give a few sentence description of what each of the output is, and why you might want to use it.
- Other small cosmetic changes and refactors

#### Is there anything you'd like reviewers to focus on?

Fixes #16476

### CODE REVIEW

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:128]: This one just outputs JSON-serialized lint results. Useful when you want to do something with it programmatically, because it's structured data.

Context: https://github.com/eslint/eslint/issues/3036

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:118]: Like JSON, but with additional data.

Context: https://github.com/eslint/rfcs/pull/10

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:186]: https://testanything.org/

When you want to pass the results to a system that expects this format.

Context: , https://github.com/eslint/eslint/issues/608

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:256]: This is another common format.

Context: https://github.com/eslint/eslint/issues/2991

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:138]: Another one that you use when you want to pass the data to a system that expects a specific format.

Context: https://github.com/eslint/eslint/issues/16

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:162]: Readable output for CLI users. This is the default formatter.

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:70]: Another one to pass the data to a system that expects a specific format,

Context: https://github.com/eslint/eslint/issues/14 (not much context there though, here's [checkstyle docs](https://checkstyle.sourceforge.io/)).

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:108]: There's not much context in https://github.com/eslint/eslint/issues/15, but I'm pretty sure this is another one you'd use if you need to pass data to a system that expects a particular format.

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:276]: Explained in https://github.com/eslint/eslint/issues/4708#issue-122267920

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:100]: For a nice visual presentation in a browser.

File: docs/src/user-guide/formatters/index.md
Comment (mdjermanovic) [inline docs/src/user-guide/formatters/index.md:80]: This one was added in the [initial commit](https://github.com/eslint/eslint/commit/a658d7b0e7d915750f18d666823d54ef2129a9af#diff-b3a58c91578d9380ce8b33e6f165445c2ec7352033f5cfca85d36930d71b8daf), so there's no context to find. I believe it's primarily intended to be readable output for CLI users, but will ask @nzakas to confirm.

File: docs/src/user-guide/formatters/index.md
Comment (nzakas) [inline docs/src/user-guide/formatters/index.md:45]: While we're at it, let's use a file extension:

```suggestion
`.eslintrc.json`:
```

File: docs/src/user-guide/formatters/index.md
Comment (nzakas) [reply docs/src/user-guide/formatters/index.md:108]: There is a [JSLint plugin for Jenkins](https://plugins.jenkins.io/jslint/) that expects a specific XML format to produce reports. This formatter type outputs data in the format that the Jenkins plugin expects. This was necessary so that people could swap out JSLint for ESLint in their Jenkins projects.

File: docs/src/user-guide/formatters/index.md
Comment (nzakas) [reply docs/src/user-guide/formatters/index.md:70]: Back in the day, a lot of Jenkins plugins expected results in Checkstyle format. It's not nearly as popular as it was but there are still a few Jenkins plugins that expect that format:

https://plugins.jenkins.io/ui/search?query=checkstyle

File: docs/src/user-guide/formatters/index.md
Comment (nzakas) [reply docs/src/user-guide/formatters/index.md:80]: This format mimics the default output from JSHint, and yes, is meant to be easier for humans to read.

File: docs/src/user-guide/formatters/index.md
Comment (nzakas) [reply docs/src/user-guide/formatters/index.md:118]: We did this to avoid breaking changes to the `json` formatter. Adding a bunch of new data in there would have broken existing consumers.

File: docs/src/user-guide/formatters/index.md
Comment (nzakas) [reply docs/src/user-guide/formatters/index.md:138]: Meant for use with the [JUnit Jenkins plugin](https://plugins.jenkins.io/junit/).

**Review by nzakas — COMMENTED** [review]
Note: This page is autogenerated. Any changes made here will go live but will be overwrriten the next time the site is generated. The evergreen changes should be made in this file:

https://github.com/eslint/eslint/blob/main/templates/formatter-examples.md.ejs

**Comment by bpmutter** [comment]
@nzakas and @mdjermanovic thank you for the context! i'll try to fill out all the descriptions with this information. will let you know in comments if need any further clarification here.

and regarding:

> Note: This page is autogenerated. Any changes made here will go live but will be overwrriten the next time the site is generated. The evergreen changes should be made in this file:
>
> https://github.com/eslint/eslint/blob/main/templates/formatter-examples.md.ejs

at first glance, i understand what's going on with the template conceptually, but am not sure how i'd add the descriptions to it. will dig in a bit, and let you know if any questions. 

if i cannot figure out how to add the descriptions to the template, i'll add them to the markdown file, so we can at least get the copy down before later adding it to the template file.

**Comment by bpmutter** [comment]
i've updated the code to include descriptions in the templates by adding a `_manifest.js` file to the source code, which contains the descriptions.

i've also:
- updated the template to include descriptions and changes that were previously made to `docs/src/user-guide/formatters/index.md`. 
- reverted the changes to `docs/src/user-guide/formatters/index.md` since the page is generated.

generate the output as expected, you can run `npm run gensite`.

i'm not convinced that using the `_manifest.js` file as is done here is the best approach for including descriptions from the source code, even if it is effective. i look forward to the maintainer team's feedback on if/how to refactor this.

File: lib/cli-engine/formatters/_manifest.js
Comment (nzakas) [inline lib/cli-engine/formatters/_manifest.js:1]: Because this is just data, can we just create a JSON file here? I'd call it `formatter-data.json` and you'd still be able to `require()` it, but would otherwise be ignored.

Comment (bpmutter) [reply lib/cli-engine/formatters/_manifest.js:1]: sgtm, i like this approach

**Comment by bpmutter** [comment]
@nzakas implemented w the file named `formatter-meta.json` to align w the file naming convention of `rules-meta.json`

**Review by nzakas — APPROVED** [review]
LGTM. Would like @mdjermanovic to review before merging.

File: lib/cli-engine/formatters/formatters-meta.json
Comment (mdjermanovic) [inline lib/cli-engine/formatters/formatters-meta.json:20]:
```suggestion
        "description": "Outputs JSON-serialized results. The `json-with-metadata` provides the same linting results as the [`json`](#json) formatter with additional metadata about the rules applied. The linting results are included in the `results` property and the rules metadata is included in the `metadata` property.\n\nAlternatively, you can use the [ESLint Node.js API](../../developer-guide/nodejs-api) to programmatically use ESLint."
```

File: lib/cli-engine/formatters/formatters-meta.json
Comment (mdjermanovic) [inline lib/cli-engine/formatters/formatters-meta.json:24]:
```suggestion
        "description": "Outputs JSON-serialized results. The `json` formatter is useful when you want to programmatically work with the CLI's linting results.\n\nAlternatively, you can use the [ESLint Node.js API](../../developer-guide/nodejs-api) to programmatically use ESLint."
```

File: templates/formatter-examples.md.ejs
Comment (mdjermanovic) [inline templates/formatter-examples.md.ejs:23]:
```suggestion
Examples of each formatter were created from linting `fullOfProblems.js` using the `.eslintrc.json` configuration shown below.
```

File: templates/formatter-examples.md.ejs
Comment (mdjermanovic) [inline templates/formatter-examples.md.ejs:55]:
```suggestion
npx eslint --format <Add formatter here> fullOfProblems.js
```

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

**Comment by bpmutter** [comment]
@mdjermanovic ready for re-review!

**Pull request merged**
